### ISSUE TITLE
Training procedure

### ISSUE THREAD

**Issue by vitruvi** [ustc-bmec/Whole-Brain-Conv#2]
thank you for sharing your model. 
can you also publish your training procedure? 
how do you handle intervals less than 27 frames?

**Comment by ustc-bmec** [comment]
We are preparing an update of the model, and I'm sorry that we wouldn't publish more until the update.
All the input data in this demo are longer than 27 frames.
If you want to apply it to intervals less than 27 frames, please modified the input channel of the 1st 1x1x1 layer according to your input data.

**Comment by vitruvi** [comment]
I have implemented the details of training procedure, as reported in your article. However, the result is not even close to the reported performance. Are there any more details for training, that you didn't mention in the article?

Do you sample equal class distributions in a single batch? 
Is there any scheduling of data partitions, such that you start training with a limited partition and increase the partition size gradually?

**Comment by ustc-bmec** [comment]
Well, we've reported details as much as we could. We will also answer the questions as clear as we could.
Did you run the same classification as the article? Could you explain how "not even close"? As I know, someone has already repeated the work by the codes on the GitHub.

We sample all the data, for the size imbalance was not so serious (< 10 folds).
The training really costs a long time. The speed limitation was not by the GPU, but the read outs of data, and a NVME SSD hard disk will help a lot.

**Comment by vitruvi** [comment]
Do you have a loss curve graph? Training kept getting stuck at local minima in my case with ReduceLRonPlateau, as reported on the article.

optimizer = optim.Adam(model.parameters(), lr=0.001)
criterion = nn.CrossEntropyLoss()
scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', factor=0.1, patience=3, verbose=False, threshold=0.0001, threshold_mode='rel', cooldown=0, min_lr=0, eps=1e-08)

Can you share your optimizer/scheduler parameters?

**Comment by boyden** [comment]
Perhaps you should know the difference between NLLoss and CrossEntropyLoss. We use log_softmax layer as the end layer. Therefore you should use NLLoss as the loss function.

**Comment by vitruvi** [comment]
Please see pytorch implementation of cross entropy loss.
https://pytorch.org/docs/master/_modules/torch/nn/functional.html#cross_entropy
"This criterion combines `log_softmax` and `nll_loss` in a single
    function"

I am still waiting for a suggestion for training parameters.

**Comment by boyden** [comment]
This is what we used.
``` python
learning_rate = 0.001
criterion = nn.NLLLoss().to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate, betas=(0.9, 0.999), eps=1e-08, weight_decay=1e-5)
scheduler = ReduceLROnPlateau(optimizer, factor = 0.1, patience = 15, mode = 'min')
```
Because we used log_softmax layer as the last layer, our criterion is `nll_loss`

**Comment by vitruvi** [comment]
Thanks, i will try with these parameters. I removed the softmax layer and used crossentropy which is identical.

**Comment by vukovicnikola** [comment]
@vitruvi were you ever able to reproduce the results of training? I've tried using the same data and code, but no luck. I've also experimented with different learning rates, as well as optimizers and a cosine annealing schedule...

**Comment by vitruvi** [comment]
@vukovicnikola I could reproduce the results of training, however I still have difficulty reproducing visualization to be close to GLM results, as they are reported.

**Comment by vukovicnikola** [comment]
@vitruvi any chance you could share a gist/repo of your training code? would be a huge help 🙏

**Comment by vitruvi** [comment]
Their parameters above are actually what is required, in addition to the code in this repo, not more. Problem is in processing HCP data the way authors want. If you include all subtasks, the performance is not impressive.

**Comment by vukovicnikola** [comment]
Interesting. Thanks! Have you tried a smaller subset of tasks, perhaps?

**Comment by vitruvi** [comment]
For each general task, authors only use 1 subtask as a "task representation". That choice leads to the reported performance. The subset is mentioned in their article.

**Comment by ustc-bmec** [comment]
> @vitruvi were you ever able to reproduce the results of training? I've tried using the same data and code, but no luck. I've also experimented with different learning rates, as well as optimizers and a cosine annealing schedule...
Thanks for your interests.
To classify the 7 tasks of HCP is not a hard task for the model, as @vitruvi has pointed out. A simpler network with a 1*1*1 layer, a few 3D CNN layers and a average pooling may also get a good performance. 
We've also pooled all the subsets of tasks together and trained the model. The model performed good in all subsets except the gambling.
One suggestion is to train the model with longer epochs, a bit longer than we used in the article.

**Comment by vukovicnikola** [comment]
@ustc-bmec Thanks for that. Surprisingly, I cannot get the model to perform well even with 200+ epochs. What I am trying now is to see if it can predict 4 (sub)tasks: language trials vs math trials vs motor (hand) trials vs social. Equal number of volumes in each task.
I have tried several things to see if things could be improved (right now max accuracy is <0.3 for a 4-way classification):
- using relu vs leaky_relu
- initializing the network with kaiming_uniform vs xavier
- max learning rates ranging from 0.000001 to 0.01.

Happy to entertain other approaches if you can suggest. Could you perhaps share your training code?
Thank you
